java/list-kml-contents/src/main/java/com/esri/arcgisruntime/sample/listkmlcontents/MainActivity.java

```diff
@@ -0,0 +1,310 @@
+/*
+ *  Copyright 2019 Esri
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.esri.arcgisruntime.sample.listkmlcontents;
+
+import java.io.ByteArrayInputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+import android.Manifest;
+import android.content.pm.PackageManager;
+import android.graphics.drawable.BitmapDrawable;
+import android.graphics.drawable.Drawable;
+import android.os.Bundle;
+import android.os.Environment;
+import android.support.annotation.NonNull;
+import android.support.v4.app.ActivityCompat;
+import android.support.v4.content.ContextCompat;
+import android.support.v7.app.AppCompatActivity;
+import android.support.v7.widget.LinearLayoutManager;
+import android.support.v7.widget.RecyclerView;
+import android.util.Log;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import com.esri.arcgisruntime.geometry.Envelope;
+import com.esri.arcgisruntime.layers.KmlLayer;
+import com.esri.arcgisruntime.loadable.LoadStatus;
+import com.esri.arcgisruntime.mapping.ArcGISScene;
+import com.esri.arcgisruntime.mapping.Basemap;
+import com.esri.arcgisruntime.mapping.Viewpoint;
+import com.esri.arcgisruntime.mapping.view.SceneView;
+import com.esri.arcgisruntime.ogc.kml.KmlContainer;
+import com.esri.arcgisruntime.ogc.kml.KmlDataset;
+import com.esri.arcgisruntime.ogc.kml.KmlDocument;
+import com.esri.arcgisruntime.ogc.kml.KmlFolder;
+import com.esri.arcgisruntime.ogc.kml.KmlGroundOverlay;
+import com.esri.arcgisruntime.ogc.kml.KmlNetworkLink;
+import com.esri.arcgisruntime.ogc.kml.KmlNode;
+import com.esri.arcgisruntime.ogc.kml.KmlPlacemark;
+import com.esri.arcgisruntime.ogc.kml.KmlScreenOverlay;
+
+public class MainActivity extends AppCompatActivity implements KmlNodeAdapter.OnItemClickListener {
+
+  private static final String TAG = MainActivity.class.getSimpleName();
+
+  private TextView mBreadcrumbTextView;
+  private RecyclerView mRecyclerView;
+  private List<KmlNode> mKmlNodeList;
+  private List<String> mKmlNodeNames;
+  private List<BitmapDrawable> mKmlNodeUxIcons;
```

Comment:
Since you're rendering the list items in an adapter, it may be better to do all the rendering in there. And instead of using a list of `BitmapDrawable` (as they could be quite resource intensive, bitmaps usually are) you might find using an `LRUCache` is more performant. https://developer.android.com/reference/android/util/LruCache

However, if you'd rather focus on usage of our API, ignore this as it's an implementation detail in Android.

Reply:
Yeah the obvious thing to do would be to pass a list of kml nodes to the adapter, but my worry was that I'd then be moving the relevant API calls to the adapter instead of in the main activity